sklearn/decomposition/tests/test_truncated_svd.py

```diff
@@ -7,98 +7,106 @@
 from sklearn.utils import check_random_state
-from sklearn.utils.testing import (assert_array_almost_equal, assert_equal,
-                                   assert_raises, assert_greater,
+from sklearn.utils.testing import (assert_raises,
                                    assert_array_less, assert_allclose)
 
 
-# Make an X that looks somewhat like a small tf-idf matrix.
-# XXX newer versions of SciPy >0.16 have scipy.sparse.rand for this.
-shape = 60, 55
-n_samples, n_features = shape
-rng = check_random_state(42)
-X = rng.randint(-100, 20, np.product(shape)).reshape(shape)
-X = sp.csr_matrix(np.maximum(X, 0), dtype=np.float64)
-X.data[:] = 1 + np.log(X.data)
-Xdense = X.A
+@pytest.fixture(scope='module')
+def X_sparse():
+    # Make an X that looks somewhat like a small tf-idf matrix.
+    # XXX newer versions of SciPy >0.16 have scipy.sparse.rand for this.
+    shape = 60, 55
+    n_samples, n_features = shape
+    rng = check_random_state(42)
+    X = rng.randint(-100, 20, np.product(shape)).reshape(shape)
+    X = sp.csr_matrix(np.maximum(X, 0), dtype=np.float64)
+    X.data[:] = 1 + np.log(X.data)
+    return X
 
 
-def test_algorithms():
+@pytest.mark.parametrize("algorithm", ['randomized'])
+def test_solvers(X_sparse, algorithm):
     svd_a = TruncatedSVD(30, algorithm="arpack")
-    svd_r = TruncatedSVD(30, algorithm="randomized", random_state=42)
+    svd = TruncatedSVD(30, algorithm=algorithm, random_state=42)
 
-    Xa = svd_a.fit_transform(X)[:, :6]
-    Xr = svd_r.fit_transform(X)[:, :6]
-    assert_array_almost_equal(Xa, Xr, decimal=5)
+    Xa = svd_a.fit_transform(X_sparse)[:, :6]
+    Xr = svd.fit_transform(X_sparse)[:, :6]
+    assert_allclose(Xa, Xr, rtol=2e-3)
 
     comp_a = np.abs(svd_a.components_)
-    comp_r = np.abs(svd_r.components_)
+    comp = np.abs(svd.components_)
     # All elements are equal, but some elements are more equal than others.
-    assert_array_almost_equal(comp_a[:9], comp_r[:9])
-    assert_array_almost_equal(comp_a[9:], comp_r[9:], decimal=2)
+    assert_allclose(comp_a[:9], comp[:9], rtol=1e-3)
+    assert_allclose(comp_a[9:], comp[9:], rtol=2e-1, atol=1e-2)
 
 
-def test_attributes():
-    for n_components in (10, 25, 41):
-        tsvd = TruncatedSVD(n_components).fit(X)
-        assert_equal(tsvd.n_components, n_components)
-        assert_equal(tsvd.components_.shape, (n_components, n_features))
+@pytest.mark.parametrize("n_components", (10, 25, 41))
+def test_attributes(n_components, X_sparse):
+    n_features = X_sparse.shape[1]
+    tsvd = TruncatedSVD(n_components).fit(X_sparse)
+    assert tsvd.n_components == n_components
+    assert tsvd.components_.shape == (n_components, n_features)
 
 
 @pytest.mark.parametrize('algorithm', ("arpack", "randomized"))
-def test_too_many_components(algorithm):
+def test_too_many_components(algorithm, X_sparse):
+    n_features = X_sparse.shape[1]
     for n_components in (n_features, n_features + 1):
```

Comment:
That's less straightforward because we need `X_sparse.shape` where `X_sparse` is a fixture (not available in the parametrize). I considered it but decided against parametrization here.